src/svm/svr.rs

```diff
@@ -0,0 +1,538 @@
+            || self.instances.len() != other.instances.len()
+        {
+            return false;
+        } else {
+            for i in 0..self.w.len() {
+                if (self.w[i] - other.w[i]).abs() > T::epsilon() {
+                    return false;
+                }
+            }
+            for i in 0..self.instances.len() {
+                if !self.instances[i].approximate_eq(&other.instances[i], T::epsilon()) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    }
+}
+
+impl<T: RealNumber, V: BaseVector<T>> SupportVector<T, V> {
+    fn new<K: Kernel<T, V>>(i: usize, x: V, y: T, eps: T, k: &K) -> SupportVector<T, V> {
+        let k_v = k.apply(&x, &x);
+        SupportVector {
+            index: i,
+            x: x,
+            grad: [eps + y, eps - y],
+            k: k_v,
+            alpha: [T::zero(), T::zero()],
+        }
+    }
+}
+
+impl<'a, T: RealNumber, M: Matrix<T>, K: Kernel<T, M::RowVector>> Optimizer<'a, T, M, K> {
+    fn optimize(
+        x: &M,
+        y: &M::RowVector,
+        kernel: &'a K,
+        parameters: &SVRParameters<T>,
+    ) -> Optimizer<'a, T, M, K> {
+        let (n, _) = x.shape();
+
+        let mut support_vectors: Vec<SupportVector<T, M::RowVector>> = Vec::with_capacity(n);
+
+        for i in 0..n {
+            support_vectors.push(SupportVector::new(
+                i,
+                x.get_row(i),
+                y.get(i),
+                parameters.eps,
+                kernel,
+            ));
+        }
+
+        Optimizer {
+            tol: parameters.tol,
+            c: parameters.c,
+            svmin: 0,
+            svmax: 0,
+            gmin: T::max_value(),
+            gmax: T::min_value(),
+            gminindex: 0,
+            gmaxindex: 0,
+            tau: T::from_f64(1e-12).unwrap(),
+            sv: support_vectors,
+            kernel: kernel,
+        }
+    }
+
+    fn minmax(&mut self) {
+        self.gmin = T::max_value();
+        self.gmax = T::min_value();
+
+        for i in 0..self.sv.len() {
+            let v = &self.sv[i];
+            let g = -v.grad[0];
+            let a = v.alpha[0];
+            if g < self.gmin && a > T::zero() {
+                self.gmin = g;
+                self.gminindex = 0;
+                self.svmin = i;
+            }
+            if g > self.gmax && a < self.c {
+                self.gmax = g;
+                self.gmaxindex = 0;
+                self.svmax = i;
+            }
+
+            let g = v.grad[1];
+            let a = v.alpha[1];
+            if g < self.gmin && a < self.c {
+                self.gmin = g;
+                self.gminindex = 1;
+                self.svmin = i;
+            }
+            if g > self.gmax && a > T::zero() {
+                self.gmax = g;
+                self.gmaxindex = 1;
+                self.svmax = i;
+            }
+        }
+    }
+
+    fn smo(mut self) -> (Vec<M::RowVector>, Vec<T>, T) {
```

Comment:
it would be helpful to have a docstring here

Reply:
I've added references and documented this method a bit. I plan to add more solid description of SVR once I am done with SVC. It will probably be done in a separate PR.